### PR TITLE
Feature/replicator dev

### PR DESCRIPTION
**Pull request by Sandychuang8**
CBL-776
CBL-447 Pending Document Ids (https://github.com/couchbase/couchbase-lite-net/pull/1206)

### CODE REVIEW

**Comment by borrrden** [comment]
PR validation is never going to work for this because it's not against a branch that also exists in EE.  To fix it you can temporarily make a branch in EE called feature/CBL-774

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs [hunk 1]
```diff
@@ -197,9 +192,12 @@ public ListenerToken AddDocumentReplicationListener([NotNull]EventHandler<Docume
...
         {
-            _statusChanged.Remove(token);
-            if(_documentEndedUpdate.Remove(token)==0)
-                Config.Options.ProgressLevel = ReplicatorProgressLevel.Overall;
+            DispatchQueue.DispatchSync(() =>
+            {
+                _statusChanged.Remove(token);
+                if (_documentEndedUpdate.Remove(token) == 0)
+                    Config.Options.ProgressLevel = ReplicatorProgressLevel.Overall;
+            });
```
Comment (borrrden) [inline src/Couchbase.Lite.Shared/API/Sync/Replicator.cs:200]: What problem did this fix?

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs [hunk 2]
```diff
@@ -223,44 +221,51 @@ public void ResetCheckpoint()
         /// </summary>
         public void Start()
         {
+            if (_disposed) {
+                throw new ObjectDisposedException(CouchbaseLiteErrorMessage.ReplicatorDisposed);
+            }
```
Comment (borrrden) [inline src/Couchbase.Lite.Shared/API/Sync/Replicator.cs:226]: Why was this moved outside?  The disposed check could now be incorrect.

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs [hunk 3]
```diff
@@ -223,44 +221,51 @@ public void ResetCheckpoint()
...
         public void Stop()
         {
             DispatchQueue.DispatchSync(() =>
             {
-                if (_stopping) {
-                    return;
-                }
-                
-                _stopping = true;
-                _reachability?.Stop();
```
Comment (borrrden) [inline src/Couchbase.Lite.Shared/API/Sync/Replicator.cs:255]: Why was this removed?  The reason it is here is so that the reachability callback does not interfere with the stopping process.

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs [hunk 4]
```diff
@@ -398,16 +393,6 @@ private void WaitPendingConflictTasks(C4ReplicatorStatus status)
             }
         }
 
-        private void ClearRepl()
-        {
-            DispatchQueue.DispatchSync(() =>
-            {
-                Native.c4repl_free(_repl);
```
Comment (borrrden) [inline src/Couchbase.Lite.Shared/API/Sync/Replicator.cs:405]: Where is this freed now?

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs [hunk 5]
```diff
@@ -223,44 +221,51 @@ public void ResetCheckpoint()
...
+                    status = new C4ReplicatorStatus {
+                        error = err,
+                        level = C4ReplicatorActivityLevel.Stopped,
+                        progress = new C4Progress()
+                    };
                 }
-
-                WriteLog.To.Sync.I(Tag, $"{this}: Starting");
-                _retryCount = 0;
-                StartInternal();
```
Comment (borrrden) [inline src/Couchbase.Lite.Shared/API/Sync/Replicator.cs:239]: I don't see where `StartInternal` gets called now....does it ever?

Comment (borrrden) [reply src/Couchbase.Lite.Shared/API/Sync/Replicator.cs:239]: Nevermind, I see that was merged into `SetupC4Replicator`

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs (hunk 4 above)
Comment (Sandychuang8) [reply src/Couchbase.Lite.Shared/API/Sync/Replicator.cs:405]: Native.c4repl_free(_repl); is now called in Dispose/Finalizing.

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs (hunk 3 above)
Comment (Sandychuang8) [reply src/Couchbase.Lite.Shared/API/Sync/Replicator.cs:255]: StopReachabilityObserver() is called when replicator is in idle or busy and stopped states. Is `c4repl_setHostReachable` taking care Reachability Observing on the LiteCore side?

Comment (borrrden) [reply src/Couchbase.Lite.Shared/API/Sync/Replicator.cs:255]: All that method does is react to the information that is provided.  The platform still needs to monitor the actual condition and inform LiteCore.  If stop is called though, the monitoring should immediately stop.

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs (hunk 4 above)
Comment (borrrden) [reply src/Couchbase.Lite.Shared/API/Sync/Replicator.cs:405]: Oh ok, that was already there!
